### PR TITLE
Update DED for Elasticsearch 6

### PR DESCRIPTION
**Pull request by HansAdema**
By popular demand (#78) it's finally here: support for Elasticsearch 6.

The following changes were necessary:

## The string field has been removed completely.
With Elasticsearch 5, the `string` field type was deprecated (#76) in favor of separate `text` and `keyword` fields. The previous auto-mapping worked because `string` was not removed yet but that needed to be changed here.

The new auto mapping defaults are now to use `text` and `keyword` field. When using these fields with Elasticsearch DSL 2.x, a `string` field will be used instead.

## Document mapping types are going to be removed.
Document mapping types and the ability to load multiple DocTypes in a single index has been removed with Elasticsearch 6, and trying to create such indexes will throw errors. Indexes created with ES <=5 still work. The tests had to be updated to accommodate this.

[More info on the Elastic blog.](https://www.elastic.co/guide/en/elasticsearch/reference/6.1/removal-of-types.html)

## The attachment field has been removed completely.
Due to the deprecation (ES 5.x) and subsequent removal (ES 6.x) of the document mapper plugin, the Attachment field is gone as well. The ingest attachment plugin should be used instead.

I implemented a simple polyfill, but I'm not familiar enough with the plugin to verify whether this will work as expected.

## DocTypeOptions has been rewritten.
The structure of `DocTypeOptions` has been rewritten for EDSL 6, so reading object properties works slightly differently. Of course, the code is still backwards compatible.

### CODE REVIEW

**Comment by HansAdema** [comment]
It looks like the CI build has failing. This is caused by a pip error, not a test failure. If you could rerun the build, it should complete successfully.

**Comment by sabricot** [comment]
Thank you @HansAdema for all your PRs, i really need to take some time to review everything.

**Comment by zhiyonghe** [comment]
hi HansAdema :
my env: python 3.6,elasticsearch (6.1.1), elasticsearch-dsl(6.1.0), and Update DED for Elasticsearch 6 ,but when I run ./manage.py search_index --rebuild ,get the follow error:
     elasticsearch_dsl.exceptions.IllegalOperation:Index object cannot have multiple types,default already set,trying to to assign doc.
what's the reason  and how to solve the issue. thanks.

**Comment by HansAdema** [comment]
@zhiyonghe Like the message says: `Index object cannot have multiple types`. Starting with Elasticsearch 6, DocTypes are deprecated and you cannot have multiple types per index. 

Please see the [Elastic blog link](https://www.elastic.co/guide/en/elasticsearch/reference/6.1/removal-of-types.html) I posted on the original PR message for more info about this.

**Comment by zhiyonghe** [comment]
I nothing to change, only follow the example to test.
models.py:
from django.db import models    
class Car(models.Model):
        name = models.CharField()
        color = models.CharField()
        description = models.TextField()
        type = models.IntegerField(choices=[
            (1, "Sedan"),
            (2, "Truck"),
            (4, "SUV"),
        ])
documents.py:
from django_elasticsearch_dsl import DocType, Index
from .models import Car

    # Name of the Elasticsearch index
car = Index('cars')
 # See Elasticsearch Indices API reference for available settings
car.settings(
        number_of_shards=1,
        number_of_replicas=0
)

@car.doc_type
 class CarDocument(DocType):
        class Meta:
            model = Car # The model associated with this DocType

            # The fields of the model you want to be indexed in Elasticsearch
            fields = [
                'name',
                'color',
                'description',
                'type',
            ]

To create and populate the Elasticsearch index and mapping use the search_index command::

    $ ./manage.py search_index --rebuild
then throws the error: Index object cannot have multiple types
check the es, there no index cars,the command isn't to create in es auto? is there some else to do ?

**Comment by HansAdema** [comment]
@zhiyonghe And you're sure you haven't defined `Index('cars')` anywhere else as well?

If you are, then I'd like to test this for myself. Can you please update the formatting of your code so I can copy and run it without having to reverse the Markdown formatting manually?

**Comment by rrmerugu** [comment]
any plans to merge this soon ?

**Comment by zhiyonghe** [comment]
@HansAdema ,I am sure that I haven't defined Index('cars') anywhere.

models.py:
from django.db import models
class Car(models.Model):
name = models.CharField()
color = models.CharField()
description = models.TextField()
type = models.IntegerField(choices=[
(1, "Sedan"),
(2, "Truck"),
(4, "SUV"),
])

documents.py:
from django_elasticsearch_dsl import DocType, Index
from .models import Car

# Name of the Elasticsearch index
car = Index('cars')
car.settings(
number_of_shards=1,
number_of_replicas=0
)

@car.doc_type
class CarDocument(DocType):
class Meta:
        model = Car # The model associated with this DocType

        # The fields of the model you want to be indexed in Elasticsearch
        fields = [
            'name',
            'color',
            'description',
            'type',
        ]

**Comment by HansAdema** [comment]
@zhiyonghe The model and document you're using looks a lot like [these fixtures](https://github.com/HansAdema/django-elasticsearch-dsl/blob/elasticsearch-6/tests/documents.py) which are used in the [integration tests](https://github.com/HansAdema/django-elasticsearch-dsl/blob/elasticsearch-6/tests/test_integration.py), which are tested against an actual Elasticsearch installation.

In short, I cannot reproduce this from my end.

Regardless, defining indexes and tacking DocType classes to them is done by Elasticsearch DSL, not this library. DED doesn't do anything meaningful with indexes which could influence this behavior (as far as I can tell). If possible could you try to strip the Django specific code from your example and ask your question with the [Elasticsearch DSL](https://github.com/elastic/elasticsearch-dsl-py) folks.

Also, for future reference, if you paste a block of source code, please wrap it in triple backticks backticks (\`\`\`) for formatting.

**Comment by hrdymchl** [comment]
+1 for merging please :)

**Comment by pmayer** [comment]
Is there any chance of merging this soon?

File: tests/documents.py [hunk 1]
```diff
@@ -50,6 +51,7 @@ class Meta:
             'launched',
             'type',
         ]
+        doc_type = 'car_document'
```
Comment (dsanders11) [inline tests/documents.py]: @HansAdema, is this change necessary? Seems redundant with the use of `@car_index.doc_type`? I've tried this PR in my own code without adding `doc_type` to `Meta` and it seems to work fine.

**Comment by dsanders11** [comment]
@HansAdema, I split out the changes from this PR that were just formatting changes (into #90) in the hopes that it might speed-up getting this merged by making it easier to review this PR, as about 10% of the line changes were just formatting ones. So that PR could be quick to merge then this one could be rebased and there'll be less to review.

**Comment by janbuerling** [comment]
I was wondering what the current situation is regarding the pull request? Is there a chance that it is going to pull soon?

**Comment by Ripcurl99983** [comment]
merge please

**Comment by sabricot** [comment]
Upon the insistence i guess i have no choice :) 
I will take a look at it before the end of the week.

**Comment by janbuerling** [comment]
@sabricot excuse me again, but do you think you have the time to dedicate yourself to this pull request today? I would be very grateful :)

**Comment by sabricot** [comment]
Its merged and released with the 0.5.0. Thanks again to @HansAdema for the great work 👍

**Comment by safwanrahman** [comment]
Possible to update the documentation?
